src/v/raft/replicate_entries_stm.cc

```diff
@@ -85,11 +85,12 @@ replicate_entries_stm::send_append_entries_request(
     vlog(_ctxlog.trace, "Sending append entries request {} to {}", req.meta, n);
 
     req.target_node_id = n;
+
+    auto opts = rpc::client_opts(append_entries_timeout());
+    opts.resource_units = ss::make_foreign<ss::lw_shared_ptr<units_t>>(_units);
+
```

Comment:
its a little confusing what the lifetime is. all these RPCs are holding alive a shared pointer to a class member so the instance of replicate_entries_stm (ie `this`) is also holding it alive. and then later you clear the class member after waiting on the dispatch semaphore to complete so its not really clear that the lifetime of the units is any different. maybe you can explain in the commit message what the new lifetime is, or get rid of class members for the request/units so that the lifetime is tied to the control flow?

Reply:
spoke offline: the lifetime of the units held by the rpc messages extend past when dispatch sem resolves